docs/Configuration.md
# Configuration

## Users and Apps

User, OIDC App, and ProxyAuth Domain management is performed in the web interface. You can view the documentation on user management on the [User Management](User-Management.md) page.

To start setting up protected applications, there are multiple authentication options available. If the application supports OIDC integration you can follow the instructions in the [OIDC Setup](OIDC-Setup.md) guide. For alternate authentication methods, you should follow the [ProxyAuth](ProxyAuth-and-Trusted-Header-SSO-Setup.md) or [LDAP Server](LDAP-Server.md) guides.

## Environment Variables

VoidAuth is configurable primarily by environment variable. For sensitive values, you may alternatively use `FILE__*` environment variables pointing to files. For example, `FILE__STORAGE_KEY=/run/secrets/storage_key` reads the `STORAGE_KEY` value from that file. Each `FILE__*` variable is mutually exclusive with its regular counterpart; you cannot have both `FILE__STORAGE_KEY` and `STORAGE_KEY` defined.

The available environment variables and their defaults are listed below:

### App Settings

| Name | Default | Description | Required | Recommended |
| :------ | :-- | :-------- | :--- | :--- |
| APP_URL | | URL of the web interface. ex. `https://auth.example.com` or `https://example.com/auth` | 🔴 | |
| TRUSTED_PROXIES | `loopback, linklocal, uniquelocal` | Sets trusted sources of sensitive HTTP headers. See documentation for how to securely set up a reverse proxy on the [Proxies and ProxyAuth](ProxyAuth-and-Trusted-Header-SSO-Setup.md#proxy-setup) page. Accepts lists of IP addresses or CIDRs, special values `loopback` `linklocal` `uniquelocal`, `true` to trust all sources, or `false` to trust none. | | ✅ |
| STORAGE_KEY | | Storage encryption key for secret values such as keys and OIDC App Client Secrets. Must be at least 32 characters long and should be randomly generated. If you do not enter one VoidAuth will recommend one to you. | 🔴 | |
| STORAGE_KEY_SECONDARY | | Secondary storage encryption key, used when rotating the primary storage encryption key. | | |
| SESSION_DOMAIN | `${APP_URL}` Base Domain | Domain of the VoidAuth Session Cookie. This is automatically set to the Base Domain of `${APP_URL}` but may be overridden here. Must be equal to or a higher level domain than `${APP_URL}` | | |
| DEFAULT_REDIRECT | `${APP_URL}` | The home/landing/app url for your domain. This is where users will be redirected upon accepting an invitation, logout, or clicking the header logo when already on the auth home page. | | ✅ |
| SIGNUP | `false` | Whether the app allows new users to self-register themselves without invitation. | | |
| SIGNUP_REQUIRES_APPROVAL | `true` | Whether new users who register themselves require approval by an admin. Setting this to `false` while `SIGNUP` is `true` enables open self-registration; use with caution! ⚠️ | | |
| EMAIL_VERIFICATION | `true` if SMTP_HOST is set, otherwise `false` | If true, users must have an email address and will get a verification email when changing their email address before it can be used. If you are using an email provider, this should probably be `true`. | | |
| MFA_REQUIRED | `false` | If set to `true`, users must use a second security factor while logging in such as an Authenticator Token or Passkey | | |
| DISABLE_FILE_UPLOADS | `false` | If set to `true`, all file uploads will be disabled throughout the application. | | |
| API_RATELIMIT  | `60` | Rate Limit for mutating (state-changing) requests per minute per IP address. Default is `60`, one per second. | | |
| ENABLE_DEBUG  | `false` | Enables debug logging. ⚠️WARNING!⚠️ Detailed user activity will be printed in the logs. | | |

### App Customization

| Name | Default | Description | Required | Recommended |
| :------ | :-- | :-------- | :--- | :--- |
| APP_TITLE | `VoidAuth` | Title that will show on the web interface, use your own brand/app/title. | | ✅ |
| APP_PORT | `3000` | The port that app will listen on. Can also be set to a unix-socket, ex. `/tmp/sock` | | |
| APP_COLOR | `#906bc7` | Theme color, rgb format; ex. #xxyyzz | | ✅ |
| APP_FONT | `monospace` | Font used in the web interface and sent emails. Safe fonts should be used, if a font is missing it will fallback to default. Multiple font families may be chosen in fallback-font format. ex. `APP_FONT: "Tahoma, Verdana, sans-serif"` | | |
| CONTACT_EMAIL | | The email address used for 'Contact' links, which are shown on most end-user pages if this is set. | | |

### Database Settings

When using the `sqlite` database adapter type, no additional database connection variables are required. You will need a mounted volume to hold the generated `db.sqlite` file, as shown in the SQLite docker compose example above.

| Name | Default | Description | Required | Recommended |
| :------ | :-- | :-------- | :--- | :--- |
| DB_ADAPTER | `postgres` | Allowed values are `postgres` and `sqlite`. | | |
| DB_HOST | | Host address of the database. May also be a postgres Unix Domain Socket, see node-postgres Unix Domain Socket documentation [here](https://node-postgres.com/features/connecting#unix-domain-sockets). | 🔴 (unless using SQLite database) | |
| DB_PASSWORD | | Password of the database. Not used if using SQLite database. | | ✅ |
| DB_PORT | `5432` | Port of the database. Not used if using SQLite database. | | |
| DB_USER | `postgres` | Username used to sign into the database by the app. Not used if using SQLite database. | | |
| DB_NAME | `postgres` | Database name used to connect to the database by the app. Not used if using SQLite database. | | |
| DB_SSL | `false` | Enables SSL connection to the database. | | |
| DB_SSL_VERIFICATION | `true` | If DB_SSL is enabled, whether to verify the SSL certificate. | | |

### Database Migration Settings

Use the following environment variables to configure a database migration. These variables *exactly* mirror the `DB_*` environment variables and describe the connection to be made to the new database. See details on how to migrate an existing database to a new one on the [Database Migration](DB-Migration.md) page.

| Name | Default | Description | Required | Recommended |
| :------ | :-- | :-------- | :--- | :--- |
| MIGRATE_TO_DB_ADAPTER | `postgres` | Allowed values are `postgres` and `sqlite`. | | |
| MIGRATE_TO_DB_HOST | | Host address of the database. | | |
| MIGRATE_TO_DB_PASSWORD | | Password of the database. Not used if migrating to SQLite database. | | ✅ |
| MIGRATE_TO_DB_PORT | `5432` | Port of the database. Not used if migrating to SQLite database. | | |
| MIGRATE_TO_DB_USER | `postgres` | Username used to sign into the database by the app. Not used if migrating to SQLite database. | | |
| MIGRATE_TO_DB_NAME | `postgres` | Database name used to connect to the database by the app. Not used if migrating to SQLite database. | | |
| MIGRATE_TO_DB_SSL | `false` | Enables SSL connection to the database. | | |
| MIGRATE_TO_DB_SSL_VERIFICATION | `true` | If MIGRATE_TO_DB_SSL is enabled, whether to verify the SSL certificate. | | |

### SMTP Settings

All of these settings are ✅ recommended to be set to the correct values for your email provider.

| Name | Default | Description |
| :------ | :-- | :-------- |
| SMTP_HOST | | SMTP Host; ex. `mail.example.com` |
| SMTP_FROM | | SMTP From **address**, should be a plain email address (ex. `app@example.com`). The SMTP From **name** will be populated with the `APP_TITLE` value. |
| SMTP_PORT | `587` | SMTP port to use. |
| SMTP_SECURE | `false` | SMTP has TLS/SSL enabled. |
| SMTP_USER | | SMTP username used to sign into email provider; ex `user@example.com` |
| SMTP_PASS | | SMTP password used to sign into email provider |
| SMTP_NOAUTH | `false` | Disable SMTP authentication. Useful for SMTP servers that do not require authentication, such as servers using IP whitelisting. When set to `true`, `SMTP_USER` and `SMTP_PASS` are not required. |
| SMTP_IGNORE_CERT | `false` | SMTP Connection will ignore invalid and self-signed certificates from SMTP providers |
| SMTP_TLS_CIPHERS | | TLS cipher to use, only set if required by your SMTP provider |
| SMTP_TLS_MIN_VERSION | | Minimum TLS version, only set if required by your SMTP provider. Possible values are `TLSv1.3`, `TLSv1.2`, `TLSv1.1`, `TLSv1` |

### LDAP Settings

LDAP is disabled by default. See the [LDAP Server](LDAP-Server.md) guide for setup details and client examples.

> [!WARNING]
> LDAP Server functionality is experimental.

| Name | Default | Description | Required | Recommended |
| :------ | :-- | :-------- | :--- | :--- |
| LDAP_ENABLED | `false` | Enables the read-only LDAP server. | | |
| LDAP_PORT | `3890` | Port the LDAP server listens on. | | |
| LDAP_BASE_DN | `dc=voidauth` | Base distinguished name for LDAP directory entries. | | |
| LDAP_BIND_DN | `cn=ldap_bind,dc=voidauth` | Service account DN LDAP clients can bind as before searching. | | |
| LDAP_BIND_PASSWORD | | Password for the LDAP service account bind DN. | 🔴 if LDAP_ENABLED is set. | |
| LDAP_TLS_CERT_FILE | | Path to a PEM certificate file. If set, `LDAP_TLS_KEY_FILE` must also be set and VoidAuth listens with LDAPS. | | |
| LDAP_TLS_KEY_FILE | | Path to the PEM private key file for `LDAP_TLS_CERT_FILE`. | | |

### Misc.

| Name | Default | Description | Required | Recommended |
| :------ | :-- | :-------- | :--- | :--- |
| PASSWORD_STRENGTH | `3` | The minimum strength of users passwords, at least 3 is recommended. Must be between 0 - 4 inclusive. | | |
| ADMIN_EMAILS | `hourly` | The minimum duration between admin notification emails. Can be set to values like: '4 hours', '30 minutes', 'weekly', 'daily', etc. or a number in seconds. If set to 'false', admin notification emails are disabled. | | | 
| DEFAULT_USER_EXPIRES_IN | | The default duration before a new users access will expire as shown on the Invitation page. Can be set to values like: '4 hours', '30 minutes', '1 week', '2 days', etc. or a number in seconds. | | | 

> [!IMPORTANT]
> Some configuration options only work when used together. `EMAIL_VERIFICATION` should only be set if the `SMTP_` options are also set. Likewise, `SIGNUP_REQUIRES_APPROVAL` does nothing unless `SIGNUP` is set.

## Config Directory

Your own branding can be applied to the app by mounting the **/app/config** directory and adding files or modifying the existing files.

The logo images of the web interface can be customized by placing your own images in the mounted **/app/config/branding** directory. The files you can add are listed below:

| Name | Extension |
| :--- | :-------- |
| logo | `svg`, `png` |
| favicon | `svg`, `png` |
| apple-touch-icon | `png` |

> [!WARNING]
> If any custom branding images are found in the mounted **/app/config/branding** directory, the default VoidAuth images will not be available. You may want to make sure that custom `logo`, `favicon`, and `apple-touch-icon` are all added together.

For information on how to change the email templates used for invitations, password resets, email verification, etc. see the documentation page for [Email Templates](Email-Templates.md).

You may also add/modify the `custom.css` file located in the **/app/config/branding** directory to add your own styling to the web interface.

## Uploaded Assets

> [!WARNING]
> For an application host, user uploaded files may require special care. Image uploads may need to be moderated by external tooling, or need privacy, compliance, or legal considerations. All file uploads can be disabled by setting the `DISABLE_FILE_UPLOADS=true` environment variable.

Uploaded application assets are stored in the **/app/config/uploads** directory, which contains any files uploaded by users. The directory structure within is based on the `SHA-256` hash of the uploaded file, and files are de-duplicated by their hashes. File names are randomized to prevent enumeration or discovery by known-hash.

## Customization

> [!IMPORTANT]
> There are enough branding options between environment variables like `APP_TITLE`, `APP_COLOR`, and config directory customization to remove any end-user reference to VoidAuth branding. You can make it your own! Below is an example of some theming changes and light mode enabled:
>
> <img width="260" src="/public/screenshots/66152d9b-b041-4374-91ec-4363ab1cb064.png" />

## Experimental

> [!WARNING]
> The following configurations are not well supported or tested, but may cover additional use-cases.

### Multi-Domain Protection

You can secure multiple domains you own while maintaining a single set of user accounts by running multiple instances of VoidAuth using the same database. They should have the same **STORAGE_KEY** and **DB_\*** variables, but may otherwise have completely different configurations. The **APP_URL** variables of each would cover a different domain. If the domains you were trying to secure were `example.com` and `your-domain.net` you might set the **APP_URL** variables like `https://auth.example.com` and `https://id.your-domain.net`. These two instances would share everything in the shared DB, including users, OIDC Apps, ProxyAuth Domains, etc.


server/util/config.ts
import { exit } from 'node:process'
import { readFileSync } from 'node:fs'
import { booleanString } from './util'
import { logger } from './logger'
import * as psl from 'psl'
import type { ClientResponse } from '@shared/api-response/ClientResponse.js'
import Docker from 'dockerode'
import { clientUpsertValidator } from '@shared/api-request/admin/ClientUpsert'
import zod from 'zod'
import type { SecureVersion } from 'node:tls'
import { randomBytes } from 'node:crypto'
import { parseDN } from '../ldap/util'
import { CLIENT_DEFAULTS } from '@shared/constants'

// basic config for app
class Config {
  APP_TITLE = 'VoidAuth'
  APP_URL = ''
  APP_PORT: number | string = 3000
  TRUSTED_PROXIES: string | boolean = 'loopback, linklocal, uniquelocal'

  SESSION_DOMAIN?: string

  SIGNUP = false
  SIGNUP_REQUIRES_APPROVAL = true
  EMAIL_VERIFICATION: boolean | null = null
  MFA_REQUIRED: boolean = false
  DISABLE_FILE_UPLOADS: boolean = false

  APP_COLOR = '#906bc7'
  APP_FONT = ''

  // Database config
  DB_ADAPTER = 'postgres'
  DB_PASSWORD?: string // checked for validity
  DB_HOST?: string
  DB_PORT?: number
  DB_USER?: string
  DB_NAME?: string
  DB_SSL: boolean = false
  DB_SSL_VERIFICATION: boolean = true

  // Database migration config
  MIGRATE_TO_DB_ADAPTER = 'postgres'
  MIGRATE_TO_DB_PASSWORD?: string
  MIGRATE_TO_DB_HOST?: string
  MIGRATE_TO_DB_PORT?: number
  MIGRATE_TO_DB_USER?: string
  MIGRATE_TO_DB_NAME?: string
  MIGRATE_TO_DB_SSL: boolean = false
  MIGRATE_TO_DB_SSL_VERIFICATION: boolean = true

  // required and checked for validity
  STORAGE_KEY: string = ''

  // Optional
  STORAGE_KEY_SECONDARY?: string
  PASSWORD_STRENGTH = 3
  API_RATELIMIT = 60
  DEFAULT_REDIRECT?: string
  CONTACT_EMAIL?: string
  ADMIN_EMAILS?: number
  DEFAULT_USER_EXPIRES_IN?: number

  // SMTP
  SMTP_HOST?: string
  SMTP_FROM?: string
  SMTP_PORT = 587
  SMTP_SECURE = false
  SMTP_USER?: string
  SMTP_PASS?: string
  SMTP_NOAUTH: boolean = false
  SMTP_IGNORE_CERT: boolean = false
  SMTP_TLS_CIPHERS?: string
  SMTP_TLS_MIN_VERSION?: SecureVersion

  // LDAP
  LDAP_ENABLED: boolean = false
  LDAP_PORT: number = 3890
  LDAP_BASE_DN = 'dc=voidauth'
  LDAP_BIND_DN = 'cn=ldap_bind,dc=voidauth'
  LDAP_BIND_PASSWORD?: string
  LDAP_TLS_CERT_FILE?: string
  LDAP_TLS_KEY_FILE?: string

  DECLARED_CLIENTS = new Map<string, ClientResponse>()
}

const appConfig = new Config()

function assignConfigValue(key: keyof Config, value: string | undefined) {
  switch (key) {
    // positive ints
    case 'SMTP_PORT':
    case 'PASSWORD_STRENGTH':
    case 'API_RATELIMIT':
    case 'LDAP_PORT':
      appConfig[key] = posInt(value) ?? appConfig[key]
      break

    // APP_PORT must be a positive integer or a non-empty string (for unix socket)
    case 'APP_PORT':
      appConfig[key] = posInt(value) ?? (stringOnly(value) || null) ?? appConfig[key]
      break

    case 'ADMIN_EMAILS':
      appConfig[key] = stringDuration(value) ?? posInt(value) ?? (booleanString(value) === false ? null : 3600) ?? appConfig[key]
      break

    case 'DEFAULT_USER_EXPIRES_IN':
      appConfig[key] = stringDuration(value) ?? posInt(value) ?? appConfig[key]
      break

    case 'TRUSTED_PROXIES':
      // Can be a boolean or a string of comma-separated IPs/CIDRs/special values (loopback, linklocal, uniquelocal)
      appConfig[key] = booleanString(value) ?? (value || null) ?? appConfig[key]
      break

    // booleans
    case 'SMTP_SECURE':
    case 'SMTP_NOAUTH':
    case 'SMTP_IGNORE_CERT':
    case 'SIGNUP':
    case 'SIGNUP_REQUIRES_APPROVAL':
    case 'MFA_REQUIRED':
    case 'DISABLE_FILE_UPLOADS':
    case 'DB_SSL':
    case 'DB_SSL_VERIFICATION':
    case 'MIGRATE_TO_DB_SSL':
    case 'MIGRATE_TO_DB_SSL_VERIFICATION':
    case 'LDAP_ENABLED':
      appConfig[key] = booleanString(value) ?? appConfig[key]
      break

    case 'EMAIL_VERIFICATION':
      appConfig[key] = booleanString(value) ?? appConfig[key]
      break

    // non default variables
    case 'SESSION_DOMAIN':
    case 'STORAGE_KEY_SECONDARY':
    case 'DEFAULT_REDIRECT':
    case 'CONTACT_EMAIL':
    case 'DB_HOST':
    case 'DB_PASSWORD':
    case 'DB_NAME':
    case 'DB_USER':
    case 'MIGRATE_TO_DB_PASSWORD':
    case 'MIGRATE_TO_DB_HOST':
    case 'MIGRATE_TO_DB_USER':
    case 'MIGRATE_TO_DB_NAME':
    case 'SMTP_HOST':
    case 'SMTP_FROM':
    case 'SMTP_USER':
    case 'SMTP_PASS':
    case 'SMTP_TLS_CIPHERS':
    case 'LDAP_BIND_PASSWORD':
    case 'LDAP_TLS_CERT_FILE':
    case 'LDAP_TLS_KEY_FILE':
      appConfig[key] = stringOnly(value) ?? appConfig[key]
      break

    case 'SMTP_TLS_MIN_VERSION':
      appConfig[key] = ['TLSv1.3', 'TLSv1.2', 'TLSv1.1', 'TLSv1'].includes(stringOnly(value) as SecureVersion)
        ? value as SecureVersion
        : appConfig[key]
      break

    // non default pos int variables
    case 'DB_PORT':
    case 'MIGRATE_TO_DB_PORT':
      appConfig[key] = posInt(value) ?? appConfig[key]
      break

    case 'DECLARED_CLIENTS':
      break

    // The default case for all string config values
    default:
      appConfig[key] = stringOnly(value) ?? appConfig[key]
      break
  }
}

async function registerDockerListener(): Promise<Docker | undefined> {
  try {
    const docker = new Docker()
    const stream = await docker.getEvents({ filters: { type: ['container'], event: ['start', 'stop', 'restart', 'die'] } })
    stream.on('data', () => {
      refreshDeclaredClients(docker)
    })
    return docker
  } catch { /* do nothing, docker not initialized */ }
}

function refreshDeclaredClients(docker: Docker | undefined) {
  // Use separate map to prevent DECLARED_CLIENTS map from being empty during a container start/stop/restart/etc.
  const clients = new Map<string, ClientResponse>()

  // Inspect docker container labels to find OIDC client configs
  if (docker !== undefined) {
    docker.listContainers(async (_err, containers) => {
      if (!containers) return

      for (const { Id } of containers) {
        const info = await docker.getContainer(Id).inspect()
        if (!info.State.Running || info.Config.Labels['voidauth.enable'] !== 'true') continue

        for (const [rawKey, value] of Object.entries(info.Config.Labels)) {
          const key = rawKey.toLowerCase()
          if (!key.startsWith('voidauth.oidc.')) continue
          const [, , client_id, variable] = key.split('.', 4)
          if (!client_id || !variable) continue
          registerClientVariable(clients, client_id, variable, value, 'label')
        }
      }
    })
  }

  // Inspect environment variables to find OIDC client configs
  Object.keys(process.env).forEach((key) => {
    const raw = key.split('_')
    if (raw.length < 3 || raw[0] !== 'OIDC') return

    const parts = [
      raw[0],
      raw[1],
      raw.slice(2).join('_'),
    ]

    const client_id = parts[1]
    const value = process.env[key]
    const variable = parts[2]
    if (client_id === undefined || variable === undefined || value === undefined) return

    registerClientVariable(clients, client_id, variable, value, 'env')
  })

  appConfig.DECLARED_CLIENTS = clients
}

function registerClientVariable(clients: Map<string, ClientResponse>,
  client_id: string,
  variable: string,
  value: string,
  source: 'env' | 'label') {
  // Skip when value is empty
  if (!value) {
    return
  }

  // Helper function to validate client variables
  const validateClientVar = <T, I extends string | string []>(input: I,
    validator: zod.ZodType<T, I | undefined | null>) => {
    const validated = validator.safeParse(input)
    if (validated.error) {
      throw new Error(zod.prettifyError(validated.error))
    }
    return validated.data
  }

  try {
    validateClientVar(client_id, clientUpsertValidator.client_id)

    let client = clients.get(client_id)
    if (!client) {
      client = {
        client_id: client_id,
        ...CLIENT_DEFAULTS,
        groups: [],
        skip_consent: false,
        declared: source,
      }
      clients.set(client_id, client)
    }

    value = value.trim()

    // Check that the values are valid, and if so use them
    switch (variable.toUpperCase()) {
      case 'CLIENT_DISPLAY_NAME':
        client.client_name = validateClientVar(value, clientUpsertValidator.client_name)
        break
      case 'CLIENT_HOMEPAGE_URL':
        client.client_uri = validateClientVar(value, clientUpsertValidator.client_uri)
        break
      case 'CLIENT_BACKCHANNEL_LOGOUT_URI':
        client.backchannel_logout_uri = validateClientVar(value, clientUpsertValidator.backchannel_logout_uri)
        break
      case 'CLIENT_LOGO_URL':
        client.logo_uri = validateClientVar(value, clientUpsertValidator.logo_uri)
        break
      case 'CLIENT_SECRET':
        client.client_secret = validateClientVar(value, clientUpsertValidator.client_secret)
        break
      case 'CLIENT_AUTH_METHOD':
        client.token_endpoint_auth_method = validateClientVar(value, clientUpsertValidator.token_endpoint_auth_method)
        break
      case 'CLIENT_GROUPS':
        client.groups = validateClientVar(value.split(',').map(v => v.trim()), clientUpsertValidator.groups)
        break
      case 'CLIENT_REDIRECT_URLS':
        client.redirect_uris = validateClientVar(value.split(',').map(v => v.trim()), clientUpsertValidator.redirect_uris)
        break
      case 'CLIENT_RESPONSE_TYPES':
        client.response_types = validateClientVar(value.split(',').map(v => v.trim()), clientUpsertValidator.response_types)
        break
      case 'CLIENT_GRANT_TYPES':
        client.grant_types = validateClientVar(value.split(',').map(v => v.trim()), clientUpsertValidator.grant_types)
        break
      case 'CLIENT_POST_LOGOUT_URLS': {
        const urls = validateClientVar(value, clientUpsertValidator.post_logout_redirect_uri)
        client.post_logout_redirect_uris = urls ? [urls] : undefined
        break
      }
      case 'CLIENT_SKIP_CONSENT':
        client.skip_consent = validateClientVar(value, zod.stringbool())
        break
    }
  } catch (e) {
    // Log error then continue. Mask secrets to avoid leaking client secrets in logs.
    const shouldMask = typeof variable === 'string' && ['SECRET', 'KEY'].some(k => variable.toUpperCase().includes(k))
    const loggedValue = shouldMask ? maskString(value) : value

    logger({
      level: 'error',
      message: 'Error processing declared client variable',
      declared_client: {
        client_id,
        source,
        variable,
        value: loggedValue,
      },
      errors: e instanceof Error ? [e] : [{ message: String(e) }],
    })
  }
}

function maskString(input: string | undefined) {
  if (!input) {
    return ''
  }
  return '***'
}

// functions to help format config
function stringOnly(value: unknown): string | null {
  if (typeof value === 'string') {
    return value
  }

  return null
}

function posInt(value: unknown): number | null {
  return typeof value === 'string' && Number.isInteger(Number.parseFloat(value)) && Number.parseInt(value) > 0
    ? Number.parseInt(value)
    : null
}

function stringDuration(durationStr: unknown) {
  if (typeof durationStr !== 'string') {
    return
  }

  // Mapping of time units to their equivalent in seconds
  const unitMap = {
    // Singular and plural forms
    minute: 60,
    hour: 3600,
    day: 86400,
    week: 604800,
    month: 2592000, // Approximating a month as 30 days
    year: 31536000, // Approximating a year as 365 days

    daily: 86400,
  } as const

  // Normalize the input string
  const normalized = durationStr.toLowerCase().trim()

  // Check for direct frequency matches
  if (Object.keys(unitMap).includes(normalized)) {
    return unitMap[normalized as keyof typeof unitMap]
  }

  // Parse numeric durations with units
  const matchNumeric = normalized.match(/(\d+)\s*(minute|hour|day|week|month|year)s?/)
  if (matchNumeric) {
    const [, numberStr, unit] = matchNumeric
    if (numberStr && unit) {
      const number = parseInt(numberStr, 10)

      if (Object.keys(unitMap).includes(unit)) {
        return number * unitMap[unit as keyof typeof unitMap]
      }
    }
  }

  // Parse unit durations directly, or ending with 'ly'
  const matchDirect = normalized.match(/(minute|hour|week|month|year)ly?/)
  if (matchDirect) {
    const [, unit] = matchDirect
    if (unit) {
      if (Object.keys(unitMap).includes(unit)) {
        return unitMap[unit as keyof typeof unitMap]
      }
    }
  }

  return
}

const configKeys = Object.getOwnPropertyNames(appConfig) as (keyof Config)[]

// read from process env vars or the given FILE__
for (const key of configKeys) {
  const envValue = process.env[key]
  const fileEnvKey = `FILE__${key}`
  const fileEnvKeyValue = process.env[fileEnvKey]

  // If both env and file env are set, log error and exit
  if (envValue && fileEnvKeyValue) {
    logger({
      level: 'error',
      message: `Environment variables ${key} and ${fileEnvKey} cannot both be set.`,
    })
    exit(1)
  } else if (fileEnvKeyValue) {
    // If FILE__* is set, must use that file value (do not fallback to regular env var)
    let fileValue: string | undefined = undefined
    try {
      fileValue = readFileSync(fileEnvKeyValue, 'utf8').trim()
    } catch (e) {
      // Log error and exit if file cannot be read
      logger({
        level: 'error',
        message: `Error reading file for environment variable ${fileEnvKey}.`,
        errors: e instanceof Error ? [e] : [{ message: String(e) }],
      })
      exit(1)
    }
    assignConfigValue(key, fileValue)
  } else {
    assignConfigValue(key, envValue)
  }
}

refreshDeclaredClients(await registerDockerListener())

/**
 * Validations and Coercions
 */

// make sure APP_URL does not have trailing slash(es)
appConfig.APP_URL = appConfig.APP_URL.replace(/\/+$/, '')

// check APP_URL is set
if (!appConfig.APP_URL || !URL.parse(appConfig.APP_URL)) {
  logger({
    level: 'error',
    message: 'APP_URL must be set and be a valid URL starting with http(s)://',
  })
  exit(1)
}

// If APP_URL hostname seems to be a private dns zone, debug log that
const pslParsedAppUrl = psl.parse(appUrl().hostname)
if ('listed' in pslParsedAppUrl && !pslParsedAppUrl.listed) {
  logger({
    level: 'debug',
    message: `APP_URL: '${appConfig.APP_URL}' appears to be in a private DNS zone.`,
  })
} else {
  logger({
    level: 'debug',
    message: `APP_URL: '${appConfig.APP_URL}'`,
  })
}

const calculatedSessionDomain = getSessionDomain()

if (calculatedSessionDomain) {
  logger({
    level: 'debug',
    message: `Session Domain: '${calculatedSessionDomain}'`,
  })
} else {
  logger({
    level: 'debug',
    message: 'Session Domain is empty, session cookies will be set for the exact host only.',
  })
}

// If SESSION_DOMAIN is set, make sure SESSION_DOMAIN cookies reach APP_URL
if (calculatedSessionDomain) {
  if (!sessionDomainReaches(appUrl().hostname)) {
    logger({
      level: 'error',
      message: 'If SESSION_DOMAIN is set, the APP_URL hostname must end with the SESSION_DOMAIN.',
    })
    exit(1)
  }
}

// check that TRUSTED_PROXIES is valid
if (typeof appConfig.TRUSTED_PROXIES === 'string') {
  // TRUSTED_PROXIES is a string, check that it is a comma-separated list of valid IPs/CIDRs/special values
  const trustedProxies = appConfig.TRUSTED_PROXIES.split(',').map(s => s.trim())
  const parsedProxies = zod.array(
    zod.union([zod.ipv4(), zod.ipv6(), zod.cidrv4(), zod.cidrv6(), zod.enum(['loopback', 'linklocal', 'uniquelocal'])]),
  ).safeParse(trustedProxies)

  if (!parsedProxies.success) {
    logger({
      level: 'error',
      message: `TRUSTED_PROXIES must be a boolean or a string of comma-separated IPs/CIDRs/special values (loopback, linklocal, uniquelocal).`,
    })
    exit(1)
  }
} else if (appConfig.TRUSTED_PROXIES) {
  logger({
    level: 'info',
    message: `TRUSTED_PROXIES is 'true', which may not be secure. Please set to the IP address or CIDR range of your proxy instead.`,
  })
} else {
  logger({
    level: 'info',
    message: `TRUSTED_PROXIES is 'false'. Rate limiting will be applied per-proxy IP address, which may cause denial of service to valid requests.`,
  })
}

// check DEFAULT_REDIRECT is valid if set
if (appConfig.DEFAULT_REDIRECT && !URL.parse(appConfig.DEFAULT_REDIRECT)) {
  logger({
    level: 'error',
    message: 'DEFAULT_REDIRECT must be a valid URL starting with http(s):// if it is set.',
  })
  exit(1)
}

// check that STORAGE_KEY is set
if (appConfig.STORAGE_KEY.length < 32) {
  logger({
    level: 'error',
    message: 'STORAGE_KEY must be set and be at least 32 characters long. Use something long and random like: '
      + randomBytes(24).toString('base64url'),
  })
  exit(1)
}

// check PASSWORD_STRENGTH is between 0 and 4
if (appConfig.PASSWORD_STRENGTH < 0 || appConfig.PASSWORD_STRENGTH > 4) {
  logger({
    level: 'error',
    message: 'PASSWORD_STRENGTH must be between 0 and 4.',
  })
  exit(1)
}

// If EMAIL_VERIFICATION is unset, give it a default value
if (appConfig.EMAIL_VERIFICATION == null) {
  appConfig.EMAIL_VERIFICATION = !!appConfig.SMTP_HOST
}

// If SMTP_NOAUTH is true, check that SMTP_USER and SMTP_PASS are not set
if (appConfig.SMTP_NOAUTH && (appConfig.SMTP_USER || appConfig.SMTP_PASS)) {
  logger({
    level: 'error',
    message: 'SMTP_USER and SMTP_PASS must not be set when SMTP_NOAUTH is true.',
  })
  exit(1)
}

// Make sure APP_FONT, if set, is in the proper format
appConfig.APP_FONT = appConfig.APP_FONT.trim()
if (appConfig.APP_FONT) {
  if (!appConfig.APP_FONT.endsWith(',')) {
    appConfig.APP_FONT += ','
  }
  appConfig.APP_FONT += ' '
}

if (appConfig.LDAP_ENABLED) {
  appConfig.LDAP_BASE_DN = appConfig.LDAP_BASE_DN.trim()

  if (!appConfig.LDAP_BASE_DN) {
    logger({
      level: 'error',
      message: 'LDAP_BASE_DN must be set and non-empty when LDAP_ENABLED is true.',
    })
    exit(1)
  }

  // Ensure correct format of LDAP_BASE_DN by trying to parse DN
  const invalidDnConfigs = [
    { key: 'LDAP_BASE_DN', value: appConfig.LDAP_BASE_DN },
  ].filter((config) => {
    try {
      return !parseDN(config.value)
    } catch {
      return true
    }
  })

  if (invalidDnConfigs[0]) {
    const firstInvalidConfig = invalidDnConfigs[0]
    logger({
      level: 'error',
      message: `Invalid LDAP DN format: ${firstInvalidConfig.key}='${firstInvalidConfig.value}'`,
    })
    exit(1)
  }

  // Check that a password is set
  if (!appConfig.LDAP_BIND_PASSWORD) {
    logger({
      level: 'error',
      message: 'LDAP_BIND_PASSWORD must be set when LDAP_ENABLED is true.',
    })
    exit(1)
  }

  if (!!appConfig.LDAP_TLS_CERT_FILE !== !!appConfig.LDAP_TLS_KEY_FILE) {
    logger({
      level: 'error',
      message: 'LDAP_TLS_CERT_FILE and LDAP_TLS_KEY_FILE must be set together.',
    })
    exit(1)
  }
}

//
// Exported Utility Functions
//

export function appUrl(): URL {
  return URL.parse(appConfig.APP_URL) as URL
}

export function basePath() {
  return appUrl().pathname.replace(/\/$/, '')
}

export function getSessionDomain(): string | undefined {
  // If empty string or 'null', return undefined so cookies will be set for the exact host only
  if (appConfig.SESSION_DOMAIN === '' || appConfig.SESSION_DOMAIN === 'null') {
    return undefined
  }
  return appConfig.SESSION_DOMAIN || getBaseDomain(appUrl().hostname)
}

export function sessionDomainReaches(hostName: string): boolean {
  const targetDomain = getBaseDomain(hostName)
  const sessionDomain = getSessionDomain()
  // Add dot to start of sessionDomain if it doesn't already have one, to prevent false positives
  const dotSD = sessionDomain && !sessionDomain.startsWith('.') ? '.' + sessionDomain : sessionDomain
  return targetDomain === sessionDomain || !!(dotSD && hostName.endsWith(dotSD))
}

//
// Utility Functions
//

function getBaseDomain(hostname: string): string | undefined {
  const isIp = zod.union([zod.ipv4(), zod.ipv6()])
  if (isIp.safeParse(hostname).success) {
    return undefined
  }

  return psl.get(hostname) ?? undefined
}

//
// Default Export
//

export default appConfig
